src/registrar/models/domain.py

```diff
@@ -1029,13 +1174,71 @@ def _fetch_cache(self, fetch_hosts=False, fetch_contacts=False):
                     cleaned["hosts"].append(
                         {k: v for k, v in host.items() if v is not ...}
                     )
-
             # replace the prior cache with new data
             self._cache = cleaned
 
         except RegistryError as e:
             logger.error(e)
 
+    def _get_or_create_public_contact(self, public_contact: PublicContact):
+        """Tries to find a PublicContact object in our DB.
+        If it can't, it'll create it."""
+        db_contact = PublicContact.objects.filter(
+            registry_id=public_contact.registry_id,
+            contact_type=public_contact.contact_type,
+            domain=self,
+        )
+
+        # Raise an error if we find duplicates.
+        # This should not occur...
+        if db_contact.count() > 1:
+            raise Exception(
+                f"Multiple contacts found for {public_contact.contact_type}"
+            )
+
+        if db_contact.count() == 1:
+            existing_contact = db_contact.get()
+            # Does the item we're grabbing match
+            # what we have in our DB?
+            # If not, we likely have a duplicate.
+            if (
+                existing_contact.email != public_contact.email
+                or existing_contact.registry_id != public_contact.registry_id
+            ):
+                raise ValueError(
+                    "Requested PublicContact is out of sync "
+                    "with DB. Potential duplicate?"
+                )
+
+            # If it already exists, we can
+            # assume that the DB instance was updated
+            # during set, so we should just use that.
+            return existing_contact
```

Comment:
This worries me that we are passing this and storing it in a list. I think it should  work given manual testing but expanding your tests below would be wise here. I will comment there

Reply:
From testing and test cases (we check on _state), I believe the list evals to store a list of references to the specified PublicContact objects, not as copies of them - that is the reasoning for this implementation. So its like in c++ how you can have a pointer and get your data from that rather than making a copy of it. When doing it this way, objects are passed by reference in Python, which is why you can do stuff like x = PublicContact.get(etc...), x.save() and have x.save() save the DB object rather than creating a duplicate DB entry